src/test/scala/org/constellation/cluster/SnapshotRollbackTest.scala

```diff
@@ -0,0 +1,70 @@
+package org.constellation.cluster
+
+import better.files.File
+import cats.implicits._
+import org.constellation.consensus.{Snapshot, SnapshotInfo, StoredSnapshot}
+import org.constellation.primitives.Transaction
+import org.constellation.serializer.KryoSerializer
+import org.scalatest.{FlatSpec, Ignore}
+
+class SnapshotRollbackTest extends FlatSpec {
+
+  private val snapshotsFolder = "rollback_data"
+  private val snapshotInfoFile = "rollback_info"
+  private val zeroSnapshotHash = Snapshot("", Seq()).hash
+
+  type AccountBalances = Map[String, Long]
+
+  "Rollback" should "validate snapshots from files" in {
+    if (File("rollback_data").notExists) cancel
+
+    val snapshots = loadSnapshotsFromFile()
+    val snapshotInfo = loadSnapshotInfoFromFile()
+
+    val newestSnapshot = findSnapshot(snapshotInfo.snapshot.lastSnapshot, snapshots)
+    val accountBalances = calculateAccountBalances(newestSnapshot.snapshot.hash, snapshots)
+
+    validateAccountBalance(accountBalances)
+  }
+
+  private def loadSnapshotsFromFile(): Seq[StoredSnapshot] =
+    File(snapshotsFolder).list.toSeq
+      .filter(_.name != snapshotInfoFile)
+      .map(s => KryoSerializer.deserializeCast[StoredSnapshot](s.byteArray))
+
+  private def loadSnapshotInfoFromFile(): SnapshotInfo =
+    KryoSerializer.deserializeCast[SnapshotInfo](File(snapshotsFolder, snapshotInfoFile).byteArray)
+
+  private def findSnapshot(hash: String, snapshots: Seq[StoredSnapshot]) =
+    snapshots.find(_.snapshot.hash == hash) match {
+      case Some(value) => value
+      case None        => throw new Exception(s"Cannot find snapshot $hash")
+    }
+
+  private def calculateAccountBalances(snapshotHash: String, snapshots: Seq[StoredSnapshot]): AccountBalances =
+    if (snapshotHash == zeroSnapshotHash) {
+      Map.empty[String, Long]
+    } else {
+      val snapshot = findSnapshot(snapshotHash, snapshots)
+      getAccountBalancesFrom(snapshot) |+| calculateAccountBalances(snapshot.snapshot.lastSnapshot, snapshots)
+    }
+
+  private def getAccountBalancesFrom(snapshot: StoredSnapshot): AccountBalances = {
```

Comment:
Why those methods are in the test file? Are those methods used for tests only?

Reply:
Only the test uses this method
I don't know where should I put method with rollback logic becasue it's strange solution to me loaded snapshots by tests, this is business logic not test